plugins/kinetic-dp/fu-kinetic-dp-firmware.c

```diff
@@ -0,0 +1,408 @@
+}
+
+guint32
+fu_kinetic_dp_firmware_get_std_fw_ver(FuKineticDpFirmware *self)
+{
+	FuKineticDpFirmwarePrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_KINETIC_DP_FIRMWARE(self), 0);
+	return priv->std_fw_ver;
+}
+
+void
+fu_kinetic_dp_firmware_set_customer_fw_ver(FuKineticDpFirmware *self, guint32 customer_fw_ver)
+{
+	FuKineticDpFirmwarePrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_KINETIC_DP_FIRMWARE(self));
+	priv->customer_fw_ver = customer_fw_ver;
+}
+
+guint32
+fu_kinetic_dp_firmware_get_customer_fw_ver(FuKineticDpFirmware *self)
+{
+	FuKineticDpFirmwarePrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_KINETIC_DP_FIRMWARE(self), 0);
+	return priv->customer_fw_ver;
+}
+
+void
+fu_kinetic_dp_firmware_set_customer_project_id(FuKineticDpFirmware *self,
+					       guint32 customer_project_id)
+{
+	FuKineticDpFirmwarePrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_KINETIC_DP_FIRMWARE(self));
+	priv->customer_project_id = customer_project_id;
+}
+
+guint8
+fu_kinetic_dp_firmware_get_customer_project_id(FuKineticDpFirmware *self)
+{
+	FuKineticDpFirmwarePrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_KINETIC_DP_FIRMWARE(self), 0);
+	return priv->customer_project_id;
+}
+
+guint32
+fu_kinetic_dp_firmware_get_valid_payload_size(const guint8 *payload_data, const guint32 data_size)
+{
+	guint32 i = 0;
+	payload_data += data_size - 1; /* start searching from the end of payload */
+	while ((*(payload_data - i) == 0xFF) && (i < data_size))
+		i++;
+	return (data_size - i);
+}
+
+static gboolean
+fu_kinetic_dp_firmware_parse(FuFirmware *self,
+			     GBytes *fw_bytes,
+			     gsize offset,
+			     FwupdInstallFlags flags,
+			     GError **error)
+{
+	FuKineticDpFirmware *firmware = FU_KINETIC_DP_FIRMWARE(self);
+	FuKineticDpFirmwarePrivate *priv = GET_PRIVATE(firmware);
+	const guint8 *buf;
+	gsize bufsz;
+	guint32 app_fw_payload_size = 0;
+	g_autoptr(GBytes) isp_drv_payload = NULL;
+	g_autoptr(GBytes) app_fw_payload = NULL;
+	g_autoptr(FuFirmware) isp_drv_img = NULL;
+	g_autoptr(FuFirmware) app_fw_img = NULL;
+
+	/* parse firmware according to Kinetic's FW image format
+	 * FW binary = 4 bytes Header(Little-Endian) + ISP driver + App FW
+	 * 4 bytes Header: size of ISP driver */
+	buf = g_bytes_get_data(fw_bytes, &bufsz);
+	if (!fu_memread_uint32_safe(buf, bufsz, 0, &priv->isp_drv_size, G_LITTLE_ENDIAN, error))
+		return FALSE;
+
+	/* app firmware payload size */
+	app_fw_payload_size =
+	    g_bytes_get_size(fw_bytes) - HEADER_LEN_ISP_DRV_SIZE - priv->isp_drv_size;
+
+	/* add ISP driver as a new image into firmware */
+	isp_drv_payload =
+	    g_bytes_new_from_bytes(fw_bytes, HEADER_LEN_ISP_DRV_SIZE, priv->isp_drv_size);
+	isp_drv_img = fu_firmware_new_from_bytes(isp_drv_payload);
+	fu_firmware_set_idx(isp_drv_img, FU_KT_FW_IMG_IDX_ISP_DRV);
+	fu_firmware_add_image(self, isp_drv_img);
+
+	/* add App FW as a new image into firmware */
+	app_fw_payload = g_bytes_new_from_bytes(fw_bytes,
```

Comment:
I think `fu_bytes_new_offset()` would be safer here

Reply:
fixed